[ADD] delivery_seur

rlizana opened the pull request:
This module adds SEUR carrier provider.
Register shipping, generate label, read shipping state and cancel shippment by webservices

This PR depends this other https://github.com/OCA/delivery-carrier/pull/266

SPANISH OCA Days 2020

pedrobaeza commented:
Puedes añadir temporalmente en oca_dependencies.txt tu remote y rama para que esto funcione mientras se fusiona el otro PR.

chienandalu commented on delivery_carrier_seur/tests/test_delivery_carrier_seur.py:
```diff
@@ -0,0 +1,114 @@
+            'delivery_type': 'seur',
+            'product_id': product_shipping_cost.id,
+            #
+            # For tests, please fill next information
```
Haz un mockup para preparar las respuestas que debería devolver SEUR, de modo que se puedan hacer el test de regresión en cualquier CI.

rlizana replied:
La prueba unitaria de sus dependencias, en este caso  (https://github.com/OCA/delivery-carrier/blob/dabc5d127546b4be5f94a703150742031cf7704d/delivery_carrier_state/tests/test_delivery_carrier_state.py) ya se encarga de probar el correcto funcionamiento de la parte de negocio en el ERP, esta prueba solo tiene sentido si tienes los datos de integración de SEUR.

chienandalu replied:
Ya, pero es que este test no hace nada :)

rlizana replied:
Aplicando el concepto TDD, si vas a continuar el desarrollo te será necesario. Pero vamos que si ves que es absurdo, no hay problema, quito los test y listo.  Ya me dices.

chienandalu replied:
Me refería a algo así:

https://github.com/OCA/l10n-spain/blob/48fd5c9834bb9ef61b12063509789e729994f13c/delivery_gls_asm/tests/test_delivery_gls_asm.py#L59-L68

chienandalu replied:
De todos modos, déjalo así y apunta en el roadmap que se puede mejorar.

chienandalu commented on delivery_carrier_seur/models/delivery_carrier.py:
```diff
@@ -0,0 +1,396 @@
+            return RewritePlugin('//*[text()="-RewritePlugin-"]', xml)
+
+        history = HistoryPlugin()
+        client = Client(
```
¿no sería mejor encapsular una clase SeurRequest() para poder abstraer mejor la capa de conexión?

rlizana replied:
fuera de este modelo no vas a usar los métodos de conexion, entonces que sentido tiene hacerlo más complejo, o dicho de otra forma, que beneficio "real" tendría encapsular estos métodos en una clase.

chienandalu replied:
Francamente, creo que es mucho más legible (separación de la parte de SEUR, de la de Odoo), mantenible (cambios futuros de API) y práctico (mucho más fácil hacer mock sobre una clase separada)

rlizana replied:
"Simple is better than complex."
Crear una clase que solo se va a llamar desde un sitio, y con una relación método a método, no ayuda en nada tenerla en otro fichero separado.
Si para que sea más "facil" aplicar futuros cambios de API hay que normalizar los parametros, como puedes ver, serían una gran cantidad de parámetros algo que complicaría en exceso la integración del modelo con la clase encapsulada y viceversa. 

Quizás es que no lo estoy viendo, pero te invito a que hagas un mockup de la clase y los métodos para que pueda ver exactamente como lo aíslas de forma elegante y práctica.

chienandalu replied:
La idea es separar en la medida de lo posible la lógica de Odoo, de la de la API correspondiente. Pero bueno, tampoco es algo bloqueante.

rlizana replied:
@chienandalu ya está separada la gestión del calculo de precio en https://github.com/OCA/delivery-carrier/pull/270

chienandalu commented on delivery_carrier_seur/__manifest__.py:
```diff
@@ -0,0 +1,41 @@
+##############################################################################
+#
+#    Copyright (C) 2020-Today Trey, Kilobytes de Soluciones <www.trey.es>
+#    Copyright (C) 2020-Today FactorLibre <www.factorlibre.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

rlizana replied:
existe alguna guideline que justifique esté cambio o es por puro gusto?

chienandalu replied:
https://github.com/OCA/maintainer-tools/commit/02df7a47c104a2a37086f7265374a1e496341018
y
https://github.com/OCA/maintainer-tools/commit/4c1afb6b77b92e1610264dfb268659d13705e2f5

rlizana replied:
Entiendo que eso propone añadir en cada fichero los autores incluso en los ficheros XML.
Realizaré los cambios para ajustarlo a estos comentarios, pero no estaría de más llevarlos a la guia de estilo para que sea normalizado.

pedrobaeza replied:
Roberto, esto se normalizó desde hace muchas versiones para quitar el tocho ese de líneas de comentario que no añaden nada, y precisamente está normalizado y con la guideline en la plantilla de ejemplo que te pone mi compañero:

https://github.com/OCA/maintainer-tools/blob/master/template/module/models/model_name.py

El principio para ello es que no es necesario copiar un trozo de la licencia, si no poner un permalink a dicha licencia. Si lo ves en todo OCA se ha hecho así (incluso en v8), y Odoo ha hecho similar.

chienandalu commented on delivery_carrier_seur/__init__.py:
```diff
@@ -0,0 +1,4 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
```suggestion
```

chienandalu commented on delivery_carrier_seur/__manifest__.py:
```diff
@@ -0,0 +1,41 @@
+{
+    'name': 'Delivery carrier SEUR',
+    'summary': 'Integrate SEUR webservice',
+    'author': 'Trey (www.trey.es), '
```
Las urls de empresa se pueden poner a nivel de créditos
```suggestion
    'author': 'Trey, '
```

rlizana replied:
Por desgracia esto es algo historico, en versiones anteriores ya usabamos este método y para no perder, ni crear otra entidad en apps.odoo.com usamos siempre ese. Incluso la empresa no es Trey, es Trey, Kilobytes de Soluciones algo que nos parece largo para ponerlo aquí, y por desgracia Trey es un nombre propio usado en USA, y la web es para saber a que Trey nos referimos.

pedrobaeza replied:
OK, esto no es problema si lo quieres así. Simplemente que sepas que esa URL no será clickable, y es conveniente que siempre sea literalmente el mismo autor para poder agrupar por dicho autor (tanto en la tienda de Odoo como la de OCA). Mi compañero se refería a que puedes poner una URL clickable en `CONTRIBUTORS.rst` de esta forma:
```rst
* `Trey <https://www.trey.es>`_:

  * Roberto Lizana
```

rlizana replied:
Es correcto @pedrobaeza , pero como decía, lo dejamos de esa forma por histórico, no para que puedan hacer clic en la URL. La intención es que sepan de que trey estamos hablando.

chienandalu commented on delivery_carrier_seur/models/__init__.py:
```diff
@@ -0,0 +1,4 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
```suggestion
```

chienandalu commented on delivery_carrier_seur/models/delivery_carrier.py:
```diff
@@ -0,0 +1,396 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

chienandalu commented on delivery_carrier_seur/tests/__init__.py:
```diff
@@ -0,0 +1,4 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
```suggestion
```

chienandalu commented on delivery_carrier_seur/tests/test_delivery_carrier_seur.py:
```diff
@@ -0,0 +1,114 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

chienandalu reviewed (changes requested):
Algunos comentarios formales sobre el código:

pedrobaeza commented:
Roberto, veo que has puesto en ambos módulos el prefijo `delivery_carrier`, pero no creo que sea necesario, y solo es necesario `delivery_`: `delivery_state` y `delivery_seur`. El repositorio se llama `delivery-carrier` porque es para incluir cosas de ambas partes, pero en este caso, siguiendo el nombre `delivery` del módulo principal, la parte de `carrier` no añade información útil, ¿no crees?

rlizana commented:
@pedrobaeza si tienes acceso a la rama queda claro que es redundante la información de **delivery_carrier** pero si estas en Odoo Apps o en la lista de módulos disponibles en tu instancia si que veo la información relevante. En ambos casos son modificaciones que afecta directamente al modelo delivery.carrier.

Por poner un ejemplo, el addon **delivery_hs_code** no tiene sentido que lleve el carrier por que no afecta al modelo *delivery.carrier* de forma directa (ni indirecta :yum:)

pedrobaeza commented:
A ver, lo que dices de las apps tampoco, porque al final para eso tienes la descripción. Pero es que esto no solo afecta al carrier, si no a todo el flujo (picking incluido). Mira de todas formas los ejemplos de enterprise, que los llaman a todos `delivery_algo`, no `delivery_carrier_algo`:

https://github.com/odoo/enterprise/tree/12.0/delivery_dhl
https://github.com/odoo/enterprise/tree/12.0/delivery_easypost
https://github.com/odoo/enterprise/tree/12.0/delivery_fedex
https://github.com/odoo/enterprise/tree/12.0/delivery_ups

rlizana commented:
Y por culpa de esa forma de hacerlo, cuando haces clic en "Install more providers" te muestra el batiburrillo de módulos de delivery. Sinceramente que lo hagan de esa forma en la enterprise no significa que sea la forma correcta. 

Pero vamos que no estoy anclado en el nombre, cambiaré ambos nombres:
- delivery_carrier_state => delivery_state
- delivery_carrier_seur => delivery_seur

pedrobaeza commented:
No, yo tampoco, si no racionalizarlo y sobre todo, **acortarlo**. Sobre lo que tú dices de "Install more providers" es muy interesante, y creo que puede deberse a la categoría del módulo, no al nombre. Si es el caso, deberíamos cambiar la categoría a todos los módulos de extensión que efectivamente no añadan providers. ¿Puedes por favor ratificar que es esto?

rlizana commented:
Lo que hace la opción de "Install more providers" es filtrar por el nombre del modulo para que empiece por *delivery_*. Iban por buen caminio pero se torcieron al nombrar los módulos de carrier :yum:

```python
    @api.multi
    def install_more_provider(self):
        return {
            'name': 'New Providers',
            'view_mode': 'kanban,form',
            'res_model': 'ir.module.module',
            'domain': [['name', 'ilike', 'delivery_']],
            'type': 'ir.actions.act_window',
            'help': _('''<p class="o_view_nocontent">
                    Buy Odoo Enterprise now to get more providers.
                </p>'''),
        }

```

pedrobaeza commented:
Ya veo, qué feo... Bueno, según eso sería el mismo problema poniendo `delivery_seur` que `delivery_carrier`, ¿no?

rlizana commented:
Si, lo mismo da que da lo mismo, lo hagas y lo pongas como lo pongas y lo hagas, el resultado no es diferente. Por eso, hago el cambio de los nombres como te he dicho

pedrobaeza commented:
Gracias!!

chienandalu commented on delivery_carrier_seur/models/delivery_carrier.py:
```diff
@@ -0,0 +1,396 @@
+                return RewritePlugin('//missing-key', '')
+            xml_root = etree.Element('root')
+            xml_exp = etree.SubElement(xml_root, 'exp')
+            for _ in range(int(data[key].get('total_bultos') or 1)):
```
Cuidado, que `_` es el método de traducción :)

rlizana replied:
cierto, ya está cambiado

javierjcf commented:
Buenas,

Creo que es necesario cambiar la dependencia de deliverir_carrier_state a delivery_state debido a que se ha renombrado este módulo:
[delivery_state](https://github.com/rlizana/delivery-carrier/commit/9bb514e894e3fa37e8573eea5cd9d30914ab0e0b)

chienandalu commented:
Sí, @rlizana trabajando en ello

chienandalu commented on delivery_carrier_seur/models/delivery_carrier.py:
```diff
@@ -0,0 +1,396 @@
+
+    def seur_rate_shipment(self, order):
+        return getattr(
+            self, '%s_rate_shipment' % self.seur_price_method)(order)
```
@rlizana Creo que estabas trabajando en un módulo para tener una base común de cálculo de costes entre carriers. ¿sigues en ello?

rlizana replied:
@chienandalu correcto, estoy extrayendo esta funcionalidad a un módulo a nuevo modulo  `delivery_price_method`

rlizana replied:
PR realizado en https://github.com/OCA/delivery-carrier/pull/270

chienandalu commented:
@rlizana , ¿seguís con esto? A nosotros nos empieza a urgir un poco. Si no vais a poder terminarlo, publica lo que tengas avanzado y podemos rematarlo nosotros.

rlizana commented:
@chienandalu si, seguimos con esto. En breve subiré todo lo que falta.

chienandalu commented:
@rlizana Si podemos ayudarte en algo para desbloquear este PR y sus dependencias, dinos.

rlizana commented:
@chienandalu puedes revisar por favor, antes de aprobar este PR es necesario aprobar https://github.com/OCA/delivery-carrier/pull/266 y https://github.com/OCA/delivery-carrier/pull/270 antes

rlizana commented:
Travis en rojo, hasta que se aprueben los PR  OCA/delivery-carrier#266 y OCA/delivery-carrier#270

pedrobaeza commented:
Justo hoy David ha empezado de vacaciones 15 días, pero lo reviso yo :smiley:

rlizana committed:
[ADD] delivery_seur

[FIX] rename views to  && [FIX] small fixes

[UPD] remove depend to delivery_price_method

[UDP] change depend stock_picking_delivery_info_computation to delivery_package_number

[FIX] work with latin characters

[UPD] test replace to SavepointCase

[UPD] test replace to setUpClass

[UPD] force carrier margin to 0

[UPD] test create picelist

[REM] set a margin value

[UPD] .rst files

chienandalu commented on delivery_seur/__manifest__.py:
```diff
@@ -0,0 +1,25 @@
+    'depends': [
+        'delivery',
+        'delivery_state',
+        'stock_picking_delivery_info_computation',
```
@rlizana hemos creado un módulo específico para añadir el campo de número de bultos sin tener que tener una dependencia tan pesada: https://github.com/OCA/delivery-carrier/pull/274 (además, es compatible con el flujo estándar de Odoo de paquetes de envío, por lo que el usuario puede elegir si hacerlo a mano o con `stock.quant.package`)

rlizana replied:
Perfecto, hago el cambio de dependencia.

javierjcf commented:
Una duda rápida:
¿Este PR debería funcionar o aun queda algo por desarrollar?. Al principio medaba un error de codificación en el XML, dado que si tienes acento en la dirección de la compañia o el cliente fallaba en el método egrees en :
`node[0].text = etree.CDATA(self.value)`
 Sin el acento va bien, y manda a seur el xml con los datos, pero siempre me devuelve "SEUR excepción: ERROR", sin ningún mensaje que describa el error. ¿Alguien ha conseguido obtener el seguimiento al validar el albarán?

rlizana commented:
Hasta que el transportista no registre la recogida no hay estado de seguimiento, por eso "SEUR excepción: ERROR".
El problema de codificación si puedes pasarme la información exacta del envio, he realizado una prueba usando `ñáéíóúç` y no he tenido problemas

javierjcf commented:
Buenas, pues haciendo debug me falla con este xml:


`<?xml version="1.0" encoding="UTF-8"?>
<root>
   <exp>
      <bulto>
         <ci>10056</ci>
         <nif>ESB70367818</nif>
         <ccc>1694</ccc>
         <servicio>031</servicio>
         <producto>002</producto>
         <cod_centro />
         <total_bultos>1</total_bultos>
         <total_kilos>2.0</total_kilos>
         <pesoBulto>2.0</pesoBulto>
         <observaciones />
         <referencia_expedicion>WH/OUT/00050</referencia_expedicion>
         <ref_bulto />
         <clavePortes>F</clavePortes>
         <clavePod />
         <claveReembolso>F</claveReembolso>
         <valorReembolso>1</valorReembolso>
         <libroControl />
         <nombre_consignatario>Javier Agrelo Bea</nombre_consignatario>
         <direccion_consignatario>Pol.Ind.Barro Meis Parcela A4.5 Naves 11-12</direccion_consignatario>
         <tipoVia_consignatario />
         <tNumVia_consignatario />
         <numVia_consignatario />
         <escalera_consignatario />
         <piso_consignatario />
         <puerta_consignatario />
         <poblacion_consignatario>Barro</poblacion_consignatario>
         <codPostal_consignatario>36692</codPostal_consignatario>
         <pais_consignatario>ES</pais_consignatario>
         <email_consignatario>javier@abj-mecanizados.com</email_consignatario>
         <telefono_consignatario>+34609132697</telefono_consignatario>
         <sms_consignatario />
         <atencion_de />
         <test_preaviso>S</test_preaviso>
         <test_reparto>S</test_reparto>
         <test_email>S</test_email>
         <test_sms>N</test_sms>
         <id_mercancia>400</id_mercancia>
         <nombre_remitente>Lupeon S.L.</nombre_remitente>
         <direccion_remitente>Pol\xc3\xadgono Porto do Molle. Nave 7A</direccion_remitente>
         <codPostal_remitente>36350</codPostal_remitente>
         <poblacion_remitente>Vigo</poblacion_remitente>
         <tipoVia_remitente />
         <eci>N</eci>
         <et>N</et>
      </bulto>
   </exp>
</root>`

En direccion remitente:
<direccion_remitente>Pol\xc3\xadgono Porto do Molle. Nave 7A</direccion_remitente>

La trace es:

`
File "/opt/odoo/buildouts/lupeon/eggs/zeep-3.4.0-py3.6.egg/zeep/plugins.py", line 32, in apply_egress
    result = plugin.egress(envelope, http_headers, operation, binding_options)
  File "/opt/odoo/buildouts/lupeon/odoo-repos/spain/delivery_seur/models/delivery_carrier.py", line 27, in egress
    node[0].text = etree.CDATA(self.value)
  File "src/lxml/etree.pyx", line 3097, in lxml.etree.CDATA.__cinit__
  File "src/lxml/apihelpers.pxi", line 1540, in lxml.etree._utf8
ValueError: All strings must be XML compatible: Unicode or ASCII, no NULL bytes or control characters
`
Si voy a la compañía y quito el acento de polígono ya no me da error.

Respecto a lo que comentas de que el transportista tiene que registrar la recogida, quizás no entienda el flujo, pero al intentar validar el albarán es cuando se manda este xml, y no deja validarlo debido a la excepción. ¿Cómo hago para que registren la recogida?, yo pensaba que mandando esto es como quedaría constancia en su  sistema de la recogida, ¿me falta algún paso previo?

rlizana commented:
@javierjcf efectivamente había un problema al enviar la información en utf-8, ya está corregido, prueba de nuevo. 

Sobre el flujo, haces el envío, registras la expedición, Seur ya tiene la orden de recogida, pero hasta que el transportista no recoja el paquete cualquier consulta de estado muestra un ERROR.

chienandalu reviewed (approved):
@rlizana Haz rebase para coger el fix del modelo 111 y junta los commits y para alante :)

pedrobaeza commented on delivery_seur/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+    'depends': [
+        'delivery',
+        'delivery_package_number',
+        'delivery_price_method',
```
Esta dependencia yo diría que no es necesaria, ¿no? Acaso SEUR no devuelve precios por su API?

chienandalu replied:
Cierto :+1:

rlizana replied:
Es necesaria, SEUR no devuelve precios por API.

rlizana commented:
Por curiosidad, @pedrobaeza en el botón de merge no tienes la opción de squash??
En este caso era necesario un rebase, y para eso no hay botón :rofl: .

pedrobaeza commented:
@rlizana con el ocabot que hacemos el merge, solo se puede hacer merge. Además, hay muchos contribuidores que quieren mantener sus commits firmados, y con un squash hecho vía GitHub, se perdería la firma, por lo que lo más habitual es pedir al propio contribuidor que haga squash.

Uno de los tests de este módulo está fallando, y es sobre los precios, pero si dices que este módulo no calcula precios, creo que ese test no haría falta aquí (ya están los tests de `delivery_price_method` para eso).

rlizana commented:
merge & squash: sabia que había un motivo, gracias por la aclaración.

Al no existir un método `seur_rate_shipment` (recuerda que se cambio por la modificación del `price_method`) se lanza una excepción cuando se llama el método de obtener precio desde el `sale.order`.

El modulo `delivery` da por hecho que todos los carrier van a tener un método para calcular el precio del envio (lo puedes ver en esta linea https://github.com/OCA/OCB/blob/12.0/addons/delivery/models/sale_order.py#L36).

Esto obliga a que cualquier carrier que no pueda calcular su precio o le creas un método que devuelva algo, o bien le metes la dependencia de `delivery_price_method`. Esta funcionalidad el mismo `delivery_price_method` no puede hacer una prueba unitaria al no tener un carrier con el que comprobar el método de calculo desde el `sale.order`.

pedrobaeza commented:
OK, gracias por la aclaración. Arregla porfa el test a ver qué ocurre y fusionamos.

rlizana commented:
Tiene que haber una incompatibilidad con algún otro modulo, o quizás sea por impuestos, o por moneda, le daré una vuelta a ver si consigo replicar y detectar que ocurre.

pedrobaeza commented:
Esto suele ocurrir por utilizar la pricelist estándar. Mira si es el caso. Si cambias a SavepointCase, cambia también el `setUp` por `setUpClass` para aprovechar las ventajas que tiene. Si no, se seguirá ejecutando el método con cada test.

rlizana commented:
Cierto, se me pasó. 
No es un problema de lista de precios, aquí no aplica, es `delivery_price_method` quien calcula el precio con la regla `fixed`.

rlizana commented:
Lo curioso que los test de travis estaban en verde antes del rebase y el squash, revisaré que he rebasado por si hay algo que pueda interferir

pedrobaeza commented:
¿Estás seguro de que no es eso? Yo te diría de crear una pricelist y asignarla en el pedido de venta para asegurar.

rlizana commented:
si no pasa los test haré la prueba creando una pricelist, pensandolo bien puede que sea por la moneda de la pricelist.

rlizana commented:
P.D.: solamente estoy seguro de que yo existo. :yum:
Solopismo (https://es.wikipedia.org/wiki/Solipsismo)

pedrobaeza commented:
También parece que el email con el que estás haciendo commits no lo tienes enganchado en tu cuenta de GitHub. Así no te va a dar atribución.

rlizana commented:
Bien visto, no se te escapa nada, gracias por el aviso, ya está corregido!

rlizana commented:
@pedrobaeza efectivamente era problema del pricelist

pedrobaeza commented on delivery_seur/readme/CONFIGURATION.rst:
```diff
@@ -0,0 +1,2 @@
+This module adds SEUR carrier provider.
```
Esto está duplicado con `DESCRIPTION.rst`. De todas formas, podrías ponerlo en español, puesto que está en OCA/l10n-spain

pedrobaeza commented on delivery_seur/readme/USAGE.rst:
```diff
@@ -0,0 +1,3 @@
+You have to set Seur carrier in the stock picking you want to ship:
+ * In the stock picking form go to *Additional Info* tab and choose Seur as carrier and the service and product code. You only be able to choose this if the state of the picking is 'Ready to Transfer.
```
Formato RST incorrecto: debe tener una línea vacía entre medias y sin indentación (lo mismo para la otra línea).

rlizana replied:
solucionado, gracias!

pedrobaeza commented:
Genial, pues vamos ya con el toque final:

/ocabot merge nobump

Gracias por la contribución!

HaraldPanten commented:
Buen aporte @rlizana
